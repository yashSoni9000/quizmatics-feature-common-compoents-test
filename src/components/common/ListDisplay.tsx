import { List, ListItem, Typography, Box, Skeleton } from "@mui/material";
import ErrorComponent from "@src/components/ErrorComponent";
import { SxProps } from "@mui/system";
import UseGenericStateHooks from "@src/hooks/UseGenericStateHook";

export interface ListDisplayData {
    id: number;
    label: string;
    value: number;
}

export type ListDisplayProps = ReturnType<
    typeof UseGenericStateHooks<ListDisplayData[]>
> & {
    title?: string;
    updateData: () => void;
    className?: string;
    listSx?: SxProps;
};

export default function ListDisplay({
    title,
    data,
    isLoading,
    error,
    refreshData,
    updateData,
    className,
    listSx,
}: ListDisplayProps) {
    return (
        <Box className={`${className} flex flex-col gap-2`}>
            <ErrorComponent
                error={error}
                isLoading={isLoading}
                refreshComponent={() => refreshData(updateData)}
            >
                {!isLoading ? (
                    <>
                        <Box>
                            <Typography fontSize="medium">{title}</Typography>
                        </Box>
                        <List
                            sx={listSx}
                            className="p-0 rounded-lg flex-1 overflow-y-scroll"
                        >
                            {data.map((item) => {
                                const { id, label, value } = item;
                                return (
                                    <ListItem
                                        key={id}
                                        className="flex justify-between border-b border-b-gray-300"
                                    >
                                        <Typography>{label}</Typography>
                                        <Typography
                                            className=" rounded-md p-1.5"
                                            sx={{
                                                backgroundColor: "primary.main",
                                                color: "primary.contrastText",
                                            }}
                                        >
                                            {value}
                                        </Typography>
                                    </ListItem>
                                );
                            })}
                        </List>
                    </>
                ) : (
                    <>
                        <Skeleton className="h-6 w-24" />
                        <Skeleton className="w-full h-full m-auto scale-[100%]" />
                    </>
                )}
            </ErrorComponent>
        </Box>
    );
}
