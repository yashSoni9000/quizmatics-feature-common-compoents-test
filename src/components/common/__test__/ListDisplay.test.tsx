import {
    fireEvent,
    render,
    renderHook,
    screen,
    within,
} from "@testing-library/react";
import "@testing-library/jest-dom";

import ListDisplay, { ListDisplayData, ListDisplayProps } from "../ListDisplay";
import UseGenericStateHooks from "@src/hooks/UseGenericStateHook";

let listDisplayCardData: ReturnType<
    typeof UseGenericStateHooks<ListDisplayData[]>
>;
let updateDataMock: ReturnType<typeof jest.fn>;
const mockData = [
    { id: 1, label: "item 1", value: 20 },
    { id: 2, label: "item 2", value: 30 },
    { id: 3, label: "item 3", value: 40 },
];

beforeEach(() => {
    updateDataMock = jest.fn();
});

jest.useFakeTimers();

beforeEach(() => {
    updateDataMock = jest.fn();
    listDisplayCardData = renderHook(() =>
        UseGenericStateHooks<ListDisplayData[]>(mockData),
    ).result.current;
});

function ListDisplayComponent({ ...props }: Partial<ListDisplayProps>) {
    return (
        <ListDisplay
            {...listDisplayCardData}
            updateData={updateDataMock}
            isLoading={false}
            {...props}
        />
    );
}

describe("List Display Test", () => {
    it("Data is rendered correctly", () => {
        const { asFragment } = render(<ListDisplayComponent />);

        const list = screen.getByRole("list");

        mockData.forEach(({ label, value }) => {
            expect(within(list).getByText(label)).toBeVisible();
            expect(within(list).getByText(value)).toBeVisible();
        });

        expect(asFragment()).toMatchSnapshot();
    });

    it("Skeletons are visible when component is loading", () => {
        render(<ListDisplayComponent isLoading={true} />);

        const skeleton = document.querySelector(".MuiSkeleton-root");
        expect(skeleton).toBeVisible();
    });

    it("Simulating from error state to non-error state", () => {
        let loading = false,
            error = true;

        function GenericListDisplayComponent() {
            return (
                <ListDisplayComponent
                    isLoading={loading}
                    error={error}
                    refreshData={refreshComponentMock}
                />
            );
        }

        function rerenderComponent() {
            rerender(<GenericListDisplayComponent />);
        }

        function refreshComponentMock() {
            error = false;
            loading = true;

            setTimeout(() => {
                loading = false;
            }, 1000);
        }

        const { rerender } = render(<GenericListDisplayComponent />);

        const retryButton = screen.getByText("Retry").parentElement;
        expect(retryButton).toBeVisible();

        fireEvent.click(retryButton!);
        rerenderComponent();

        const skeleton = document.querySelector(".MuiSkeleton-root");
        expect(skeleton).toBeVisible();

        jest.advanceTimersByTime(1000);
        rerenderComponent();

        const list = screen.getByRole("list");
        const listItems = within(list).getAllByRole("listitem");

        expect(listItems).toHaveLength(mockData.length);
    });
});
