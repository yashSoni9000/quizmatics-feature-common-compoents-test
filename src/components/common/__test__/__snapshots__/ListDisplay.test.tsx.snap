// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List Display Test Data is rendered correctly 1`] = `
<DocumentFragment>
  <div
    class="undefined flex flex-col gap-2 MuiBox-root css-0"
  >
    <div
      class="MuiBox-root css-0"
    >
      <p
        class="MuiTypography-root MuiTypography-body1 css-18a2bkc-MuiTypography-root"
      />
    </div>
    <ul
      class="MuiList-root MuiList-padding p-0 rounded-lg flex-1 overflow-y-scroll css-h4y409-MuiList-root"
    >
      <li
        class="MuiListItem-root MuiListItem-gutters MuiListItem-padding flex justify-between border-b border-b-gray-300 css-1p823my-MuiListItem-root"
      >
        <p
          class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
        >
          item 1
        </p>
        <p
          class="MuiTypography-root MuiTypography-body1  rounded-md p-1.5 css-1eeia4w-MuiTypography-root"
        >
          20
        </p>
      </li>
      <li
        class="MuiListItem-root MuiListItem-gutters MuiListItem-padding flex justify-between border-b border-b-gray-300 css-1p823my-MuiListItem-root"
      >
        <p
          class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
        >
          item 2
        </p>
        <p
          class="MuiTypography-root MuiTypography-body1  rounded-md p-1.5 css-1eeia4w-MuiTypography-root"
        >
          30
        </p>
      </li>
      <li
        class="MuiListItem-root MuiListItem-gutters MuiListItem-padding flex justify-between border-b border-b-gray-300 css-1p823my-MuiListItem-root"
      >
        <p
          class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
        >
          item 3
        </p>
        <p
          class="MuiTypography-root MuiTypography-body1  rounded-md p-1.5 css-1eeia4w-MuiTypography-root"
        >
          40
        </p>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;
